### PR TITLE
ARROW-8164: [C++][Dataset] Provide Dataset::ReplaceSchema()

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

File: cpp/src/arrow/dataset/projector.cc [hunk 1]
```diff
@@ -34,6 +34,30 @@
 namespace arrow {
 namespace dataset {
 
+Status CheckProjectable(const Schema& from, const Schema& to) {
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/projector.cc]: This method should have a unit test.

Comment (bkietz) [reply cpp/src/arrow/dataset/projector.cc]: will do

File: cpp/src/arrow/result.h [hunk 2]
```diff
@@ -45,6 +45,12 @@ ARROW_EXPORT void InvalidValueOrDie(const Status& st);
 
 }  // namespace internal
 
+#if defined(__clang__)
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/result.h]: It *will* raise a warning when a `Result<>` is unused. The attribute is also attached to Status

**Review by fsaintjacques — APPROVED** [review]

**Comment by fsaintjacques** [comment]
Can you rebase with master, there's a lot of failing test noise.

**Comment by bkietz** [comment]
CI failures are [s3fs](https://github.com/apache/arrow/pull/6721/checks?check_run_id=549362748#step:6:1759), [grpc build failure](https://github.com/apache/arrow/pull/6721/checks?check_run_id=549362914#step:7:374) [x2](https://github.com/apache/arrow/pull/6721/checks?check_run_id=549362922#step:7:374) [x3](https://github.com/apache/arrow/pull/6721/checks?check_run_id=549362840#step:7:362), and [RTools flake](https://github.com/apache/arrow/pull/6732/checks?check_run_id=549606188#step:6:396)

**Comment by nealrichardson** [comment]
It's not clear that the R Windows failure is a flake, it reads similar to https://issues.apache.org/jira/browse/ARROW-8217

**Comment by jorisvandenbossche** [comment]
There is still a R failure (don't know whether this is related), and windows is failing in cmake (related to googletest)

File: cpp/src/arrow/dataset/projector.cc [hunk 3]
```diff
@@ -86,32 +113,23 @@ Result<std::shared_ptr<RecordBatch>> RecordBatchProjector::Project(
 
 Status RecordBatchProjector::SetInputSchema(std::shared_ptr<Schema> from,
                                             MemoryPool* pool) {
+  RETURN_NOT_OK(CheckProjectable(*from, *to_));
```
Comment (nealrichardson) [inline cpp/src/arrow/dataset/projector.cc:116]: If I read this diff correctly, the only place that the R package tests should touch new code (i.e. without calling ReplaceSchema, which has a binding added but no test) is here. Is there something suspicious about CheckProjectable (on 32-bit mingw)?

**Comment by nealrichardson** [comment]
R is passing on master, so we need to figure out why it's failing here.

**Comment by nealrichardson** [comment]
Reproduced the crash locally. https://github.com/apache/arrow/blob/master/r/tests/testthat/test-json.R#L30-L32 is where it exits. Can't get a backtrace with gdb, even with a debug build.

**Comment by nealrichardson** [comment]
@bkietz Here's an even shorter repro, reduced from the test that blows up:

```r
library(arrow)
tf <- tempfile()
writeLines('{ "hello": 3.5 }\n', tf, useBytes = TRUE)
read_json_arrow(tf, as_data_frame = FALSE)
```

**Comment by nealrichardson** [comment]
I did some more interactive exploration of this in R, and found that the crash happens when calling `arrow::json::TableReader`'s `Read()` method.

**Comment by nealrichardson** [comment]
🎉 @bkietz the Rtools build is passing now! But now 6 other things are failing. Should check if they're flakes or if we broke something else in this flailing.

**Comment by nealrichardson** [comment]
merging 🤞

**Pull request closed without merging**